drivers/soundwire/intel_master.c

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2020 Intel Corporation.
+
+#include <linux/device.h>
+#include <linux/acpi.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#if !IS_ENABLED(CONFIG_VIRTUAL_BUS)
+#include <linux/platform_device.h>
+#else
+#include <linux/virtual_bus.h>
+#endif
+#include "cadence_master.h"
+#include "intel_master.h"
+#include "intel.h"
+#include "bus.h"
+
+/*
+ * this driver is an abstraction of the virtbus_device or
+ * platform_device, with additional data structures needed to help
+ * manage the Intel SoundWire IP and completion structure to force a
+ * synchronous behavior so that the bus enumeration is complete before
+ * the parent can continue its initialization. This is needed e.g. to
+ * detect the platform configuration and identify the relevant ASoC
+ * machine driver.
+ */
+#if IS_ENABLED(CONFIG_VIRTUAL_BUS)
+
+static void sdw_intel_master_release(struct virtbus_device *_dev)
+{
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(_dev);
+
+	kfree(master_dev);
+}
+
+#else
+
+static void sdw_intel_master_release(struct device *_dev)
+{
+	struct platform_device *pdev = container_of(_dev, struct platform_device, dev);
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(pdev);
```

Comment:
wouldn't you need to copy all the actions from `platform_device_release()` except freeing the device itself here too?